### PR TITLE
Updated ProFormA-XML import

### PR DESCRIPTION
**Pull request by Adrian-St**

### CODE REVIEW

File: app/models/exercise.rb [hunk 1]
```diff
@@ -300,35 +366,108 @@ def determine_file_role_from_proforma_file(task_node, file_node)
...
+  def build_proforma_xml_for_test(xml, test, index)
+    proforma = xml['p']
+    proforma.test('id' => 't' + index.to_s) {
+      proforma.title('')
+      proforma.send('test-type', 'unittest')
+      proforma.send('test-configuration') {
+        proforma.filerefs {
+          proforma.fileref('refid' => test.id.to_s)
+        }
+        xml['u'].unittest('framework' => 'JUnit 4', 'version' => '')
```
Comment (tstaubitz) [inline app/models/exercise.rb]: You need to check here for which programming language the exercise is. If it is Java, then use JUnit 4 (split frame work and version), if it is Python then use Pyunit, if it is Ruby, then use RSpec. We intentionally decided to leave this hard-coded in here as everything else would require very high efforts, which do not bring us any further. So the check for the programming language is not a nice solution but the minimum viable product.

File: app/models/exercise.rb [hunk 2]
```diff
@@ -300,35 +366,108 @@ def determine_file_role_from_proforma_file(task_node, file_node)
...
+        proforma.fileref('refid' => model_solution_file.id.to_s)
+      }
     }
-    self.execution_environment_id = 1
+  end
+
+  def to_proforma_xml
+    builder = Nokogiri::XML::Builder.new do |xml|
+      proforma = xml['p']
+      proforma.task('xmlns:p' => 'urn:proforma:task:v1.1', 'lang' => 'en', 'uuid' => SecureRandom.uuid,
```
Comment (tstaubitz) [inline app/models/exercise.rb]: change the default language here to German, please. Most of our exercises are German, currently, so it will be less work to adjust the others.

File: app/policies/exercise_policy.rb [hunk 3]
```diff
@@ -12,7 +12,7 @@ def show?
     @user.internal_user?
   end
 
-  [:clone?, :destroy?, :edit?, :statistics?, :update?, :feedback?].each do |action|
```
Comment (tstaubitz) [inline app/policies/exercise_policy.rb]: Why did you delete the feedback here? We suspect this went wrong during the merge?

File: app/controllers/exercises_controller.rb [hunk 4]
```diff
@@ -93,6 +97,17 @@ def edit
     collect_set_and_unset_exercise_tags
   end
 
+  def push_proforma_xml
+    codeharbor_link = CodeHarborLink.find(params[:account_link])
+    oauth2Client = OAuth2::Client.new('client_id', 'client_secret', :site => codeharbor_link.push_url)
```
Comment (tstaubitz) [inline app/controllers/exercises_controller.rb]: client_id and client_secret will be set correctly in the next merge_request.
We will adjust the procedure on CodeHarbor accordingly. 
A group admin will be able to register an autograder with CodeHarbor.
The groups on CodeHarbor will be extended with a third role: instructor

- admins can add and edit users, exercises and autograders to the group.  
- members can view the exercises and export them to xml
- instructors can do the same as members plus pushing them to the registered autograder.

**Comment by tstaubitz** [comment]
@rteusner We just decided that there will be no automatic mapping from programming language (CodeHarbor) to execution environment (CodeOcean). There are too many options. Basically the execution_environment needs to be manually selected.

The problem is however, that we need an execution_environment when an exercise is created.
Solution 1: we make the execution_environment optional   
Solution 2: we add a default execution_environment  in the database with some sort of basic docker container that actually does nothing as a placeholder until the user selects a proper environment. 

A hint will be shown to the teacher either way that some sort of action is required both in the exercise list and the exercise show/edit view.

My preference would be Solution 1 as in the end the outcome is the same and the solution is less error prone and way more straight forward. 

Opinions?

File: app/views/exercises/show.html.slim [hunk 5]
```diff
@@ -7,10 +7,19 @@ h1
...
 
+
 = row(label: 'exercise.title', value: @exercise.title)
 = row(label: 'exercise.user', value: link_to_if(policy(@exercise.author).show?, @exercise.author, @exercise.author))
 = row(label: 'exercise.description', value: render_markdown(@exercise.description))
-= row(label: 'exercise.execution_environment', value: link_to_if(policy(@exercise.execution_environment).show?, @exercise.execution_environment, @exercise.execution_environment))
+= row(label: 'exercise.execution_environment') do
+  = link_to_if(policy(@exercise.execution_environment).show?, @exercise.execution_environment, @exercise.execution_environment)
+  - if @exercise.default_execution_environment
+    .default-execution-environment-warning
```
Comment (rteusner) [inline app/views/exercises/show.html.slim]: default execution environment warning? Is that specifically for imported exercises?

File: config/docker.yml.erb.travis [hunk 6]
```diff
@@ -7,7 +7,7 @@ default: &default
 
 development:
   <<: *default
-  host: tcp://127.0.0.1:2376
+  host: tcp://0.0.0.0:2376
```
Comment (rteusner) [inline config/docker.yml.erb.travis]: is this necessary? Travis seems to be working atm, why change it?

File: db/migrate/20180312135814_add_default_execution_environment_to_exercises.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddDefaultExecutionEnvironmentToExercises < ActiveRecord::Migration
+  def change
+    add_column :exercises, :default_execution_environment, :boolean, :default => false
```
Comment (rteusner) [inline db/migrate/20180312135814_add_default_execution_environment_to_exercises.rb]: how is that intended to work?

File: db/schema.rb [hunk 8]
```diff
@@ -123,18 +124,19 @@
...
-    t.string   "title",                    limit: 255
+    t.string   "title",                         limit: 255
     t.datetime "created_at"
     t.datetime "updated_at"
     t.integer  "user_id"
     t.text     "instructions"
     t.boolean  "public"
-    t.string   "user_type",                limit: 255
-    t.string   "token",                    limit: 255
+    t.string   "user_type",                     limit: 255
```
Comment (rteusner) [inline db/schema.rb]: clear those changes please...

File: debian_installer/setup_debian_1_install_postgres.sh [hunk 9]
```diff
@@ -11,7 +11,7 @@ then
   echo "Add Postgres sources..."
   cd /etc/apt/sources.list.d
   sudo touch pgdg.list
```
Comment (rteusner) [inline debian_installer/setup_debian_1_install_postgres.sh]: these changes do not belong into this pull request.

File: lib/proforma/xml_generator.rb [hunk 10]
```diff
@@ -0,0 +1,142 @@
+require 'nokogiri'
```
Comment (rteusner) [inline lib/proforma/xml_generator.rb]: this serves as the "exporter" ?

File: db/migrate/20180312135814_add_default_execution_environment_to_exercises.rb (hunk 7 above)
Comment (rteusner) [reply db/migrate/20180312135814_add_default_execution_environment_to_exercises.rb]: I kinda dislike that solution...

File: app/assets/javascripts/execution_environments.js [hunk 11]
```diff
@@ -1,7 +1,137 @@
...
   if ($.isController('execution_environments')) {
     if ($('.edit_execution_environment, .new_execution_environment').isPresent()) {
       new MarkdownEditor('#execution_environment_help');
+      setup_programming_language();
+      //show_programming_language_values();
+      setup_chosen_fields();
     }
   }
 });
+
```
Comment (rteusner) [inline app/assets/javascripts/execution_environments.js]: this is called also outside of the view / controller of execution_environments.... I suppose this is not intended?

If it is - why?

File: app/controllers/exercises_controller.rb (hunk 4 above)
Comment (rteusner) [reply app/controllers/exercises_controller.rb]: this has not yet happened.

File: app/policies/exercise_policy.rb (hunk 3 above)
Comment (rteusner) [reply app/policies/exercise_policy.rb]: We still need an answer here

File: app/policies/exercise_policy.rb [hunk 12]
```diff
@@ -12,7 +12,7 @@ def show?
...
+  [:clone?, :destroy?, :edit?, :statistics?, :update?, :push_proforma_xml?, :feedback?].each do |action|
```
Comment (rteusner) [inline app/policies/exercise_policy.rb]: ah it was re-added here...

File: app/policies/programming_language_policy.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class ProgrammingLanguagePolicy < AdminOrAuthorPolicy
+  [:create?, :versions?].each do |action|
```
Comment (rteusner) [inline app/policies/programming_language_policy.rb]: please unnest.

File: app/views/exercises/index.html.slim [hunk 14]
```diff
@@ -28,7 +28,15 @@ h1 = Exercise.model_name.human(count: 2)
       - @exercises.each do |exercise|
         tr data-id=exercise.id
           td = exercise.title
-          td = link_to_if(exercise.execution_environment && policy(exercise.execution_environment).show?, exercise.execution_environment, exercise.execution_environment)
+          td
+            - if exercise.execution_environment
+              = link_to_if(policy(exercise.execution_environment).show?, exercise.execution_environment, exercise.execution_environment)
+            - else
+              i.fa.fa-warning style='color: zred;'
```
Comment (Parnswir) [inline app/views/exercises/index.html.slim]: Avoid inline styles. Also: typo?

File: config/locales/de.yml [hunk 15]
```diff
@@ -620,6 +648,7 @@ de:
     upload_file: Datei hochladen
     time:
       before: "%{time} zuvor"
+    "yes": Ja
```
Comment (rteusner) [inline config/locales/de.yml]: Does this need to be in quotation marks?

File: app/views/exercises/show.html.slim [hunk 16]
```diff
@@ -7,10 +7,20 @@ h1
   - if policy(@exercise).edit?
     = render('shared/edit_button', object: @exercise)
 
+
 = row(label: 'exercise.title', value: @exercise.title)
 = row(label: 'exercise.user', value: link_to_if(policy(@exercise.author).show?, @exercise.author, @exercise.author))
 = row(label: 'exercise.description', value: render_markdown(@exercise.description))
-= row(label: 'exercise.execution_environment', value: link_to_if(policy(@exercise.execution_environment).show?, @exercise.execution_environment, @exercise.execution_environment))
+= row(label: 'exercise.execution_environment') do
+  - if @exercise.execution_environment
```
Comment (Parnswir) [inline app/views/exercises/show.html.slim]: This block looks nearly identical to https://github.com/openHPI/codeocean/pull/132/files#diff-65b4469e16350935fe943bda56060ae2R32 -- it might be good to extract the template and `render` it here as a component

File: lib/tasks/export_exercise_collection.rake [hunk 17]
```diff
@@ -0,0 +1,19 @@
+namespace :codeharbor do
+
+  desc 'write displaynames retrieved from the account service as csv into the codeocean database'
```
Comment (rteusner) [inline lib/tasks/export_exercise_collection.rake]: this description seems to be the leftover of a copy and paste...

File: lib/tasks/export_exercise_collection.rake [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+  task :export_exercise_collection, [:exercise_collection_id, :push_url, :push_token] => [:environment] do |t, args|
+
+    begin
+      ExerciseCollection.find(args.exercise_collection_id).exercises.each do |exercise|
+        oauth2Client = OAuth2::Client.new('client_id', 'client_secret', :site => args.push_url)
```
Comment (rteusner) [inline lib/tasks/export_exercise_collection.rake]: still hard coded strings here

File: test/models/programming_language_test.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (rteusner) [inline test/models/programming_language_test.rb]: this seems to be leftover? --> Remove

File: test/models/programming_languages_join_test.rb [hunk 20]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (rteusner) [inline test/models/programming_languages_join_test.rb]: Same here: Leftover? -> Remove.

File: app/views/execution_environments/_programming_languages_join_fields.html.slim [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+  = f.hidden_field :_destroy
+  - f.object.build_programming_language unless f.object.programming_language
+  = f.fields_for :programming_language do |ff|
+    .col-md-4.name
+      = ff.object.name
+    .col-md-4.version
+      = ff.object.version
+    .col-md-3.default
+      span.true style="display:#{display_true}"
+        = t('shared.yes')
```
Comment (rteusner) [inline app/views/execution_environments/_programming_languages_join_fields.html.slim]: please use icons here. Remove the shared strings yes and no.

File: app/views/execution_environments/_programming_languages_join_fields.html.slim [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+      span.false style="display:#{display_false}"
+        = t('shared.no')
```
Comment (rteusner) [inline app/views/execution_environments/_programming_languages_join_fields.html.slim]: same here

File: app/assets/javascripts/execution_environments.js (hunk 11 above)
Comment (Adrian-St) [reply app/assets/javascripts/execution_environments.js]: I don't really understand what you mean.
setup_programming_language() and setup_chosen_fields() is only supposed to be called when loading the form.
But I thought that's the point of these lines: 

if ($.isController('execution_environments')) {
    if ($('.edit_execution_environment, .new_execution_environment').isPresent())

Am I mistaken? If so, how do I make sure it is only called with the form?

File: lib/proforma/xml_generator.rb (hunk 10 above)
Comment (Adrian-St) [reply lib/proforma/xml_generator.rb]: Yes, nokogiri is a gem to generate and read xml data.

File: app/assets/javascripts/execution_environments.js (hunk 11 above)
Comment (Parnswir) [reply app/assets/javascripts/execution_environments.js]: I think the comment refers to the block below, lines 95-137. 😄

File: config/locales/de.yml (hunk 15 above)
Comment (Parnswir) [reply config/locales/de.yml]: `yes` and `no` are key words in YAML, they need to be in quotes, because they would otherwise evaluate to `true` and `false` [at runtime](https://makandracards.com/makandra/24809-yaml-keys-like-yes-or-no-evaluate-to-true-and-false).

File: app/assets/javascripts/execution_environments.js (hunk 11 above)
Comment (Adrian-St) [reply app/assets/javascripts/execution_environments.js]: Yeah, I saw it now, has been fixed :D

File: lib/tasks/export_exercise_collection.rake (hunk 18 above)
Comment (Adrian-St) [reply lib/tasks/export_exercise_collection.rake]: I think they are okay here, since we only use rake tasks internally and client_id and client_secret are only there because they are required by OAuth2, but are not used otherwise.
What do you think?

**Comment by MrSerth** [comment]
superseded by #423
